examples/ex3p.cpp

```diff
@@ -156,57 +174,73 @@ int main(int argc, char *argv[])
    b->AddDomainIntegrator(new VectorFEDomainLFIntegrator(f));
    b->Assemble();
 
-   // 9. Define the solution vector x as a parallel finite element grid function
-   //    corresponding to fespace. Initialize x by projecting the exact
-   //    solution. Note that only values from the boundary edges will be used
-   //    when eliminating the non-homogeneous boundary condition to modify the
-   //    r.h.s. vector b.
+   // 10. Define the solution vector x as a parallel finite element grid function
+   //     corresponding to fespace. Initialize x by projecting the exact
+   //     solution. Note that only values from the boundary edges will be used
+   //     when eliminating the non-homogeneous boundary condition to modify the
+   //     r.h.s. vector b.
    ParGridFunction x(fespace);
    VectorFunctionCoefficient E(sdim, E_exact);
    x.ProjectCoefficient(E);
 
-   // 10. Set up the parallel bilinear form corresponding to the EM diffusion
+   // 11. Set up the parallel bilinear form corresponding to the EM diffusion
    //     operator curl muinv curl + sigma I, by adding the curl-curl and the
    //     mass domain integrators.
    Coefficient *muinv = new ConstantCoefficient(1.0);
    Coefficient *sigma = new ConstantCoefficient(1.0);
    ParBilinearForm *a = new ParBilinearForm(fespace);
+   if (pa) { a->SetAssemblyLevel(AssemblyLevel::PARTIAL); }
    a->AddDomainIntegrator(new CurlCurlIntegrator(*muinv));
    a->AddDomainIntegrator(new VectorFEMassIntegrator(*sigma));
 
-   // 11. Assemble the parallel bilinear form and the corresponding linear
+   // 12. Assemble the parallel bilinear form and the corresponding linear
    //     system, applying any necessary transformations such as: parallel
    //     assembly, eliminating boundary conditions, applying conforming
    //     constraints for non-conforming AMR, static condensation, etc.
    if (static_cond) { a->EnableStaticCondensation(); }
    a->Assemble();
 
-   HypreParMatrix A;
+   OperatorPtr A;
    Vector B, X;
    a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
 
-   if (myid == 0)
+   // 13. Define and apply a parallel PCG solver for AX=B with the AMS
+   //     preconditioner from hypre, in the full assembly case.
+   //     With partial assembly, use no preconditioner, for now.
+
+   if (pa)
+   {
+      Vector tdiag_pa(fespace->GetTrueVSize());
+      a->AssembleDiagonal(tdiag_pa);
+
+      OperatorJacobiSmoother Jacobi(tdiag_pa, ess_tdof_list, 1.0);
+
+      CGSolver cg(MPI_COMM_WORLD);
+      cg.SetRelTol(1e-12);
+      cg.SetMaxIter(1000);
+      cg.SetPrintLevel(1);
+      cg.SetOperator(*A);
+      cg.SetPreconditioner(Jacobi);
+      cg.Mult(B, X);
+   }
+   else
    {
-      cout << "Size of linear system: " << A.GetGlobalNumRows() << endl;
```

Comment:
I noticed in the internal tests that this line is now missing. Can we add it back, at least for the full assembly case?